datafusion/physical-expr/src/functions.rs

```diff
@@ -45,142 +45,233 @@ use arrow::{
     datatypes::{DataType, Int32Type, Int64Type, Schema},
 };
 use datafusion_common::{internal_err, DataFusionError, Result, ScalarValue};
-pub use datafusion_expr::FuncMonotonicity;
 use datafusion_expr::{
-    BuiltinScalarFunction, ColumnarValue, ScalarFunctionImplementation,
+    BuiltinScalarFunction, ColumnarValue, FunctionReturnType,
+    ScalarFunctionImplementation,
 };
+pub use datafusion_expr::{
+    FuncMonotonicity, ReturnTypeFunction, ScalarFunctionDef, TypeSignature, Volatility,
+};
+use std::any::Any;
 use std::ops::Neg;
 use std::sync::Arc;
 
-/// Create a physical (function) expression.
-/// This function errors when `args`' can't be coerced to a valid argument type of the function.
-pub fn create_physical_expr(
-    fun: &BuiltinScalarFunction,
-    input_phy_exprs: &[Arc<dyn PhysicalExpr>],
-    input_schema: &Schema,
-    execution_props: &ExecutionProps,
-) -> Result<Arc<dyn PhysicalExpr>> {
-    let input_expr_types = input_phy_exprs
-        .iter()
-        .map(|e| e.data_type(input_schema))
-        .collect::<Result<Vec<_>>>()?;
+/// `ScalarFunctionDef` is the new interface for builtin scalar functions
+/// This is an adapter between the old and new interface, to use the new interface
+/// for internal execution. Functions are planned to move into new interface gradually
+#[derive(Debug, Clone)]
+pub(crate) struct BuiltinScalarFunctionWrapper {
+    func: Arc<dyn ScalarFunctionDef>,
+    // functions like `now()` requires per-execution properties
+    execution_props: ExecutionProps,
```

Comment:
Few functions like `now()` require the core to pass some information to it, when migrating those functions, we can extend the `trait ScalarFunctionDef` with a optional method `set_execution_props(exec_props: ExecutionProps)`, as the mechanism to let core pass data to functions defined outside the core